pep-0635.rst

```diff
@@ -0,0 +1,689 @@
+Nonetheless, we find that Python code is often sparkled with
+conditions depending on ``isinstance``, ``len``, ``getattr``,
+``hasattr``, etc.  Despite the benefits of 'duck typing', actual
+code still requires to query the format or type of an object in order
+to choose an appropriate processing action or to extract relevant bits
+of information.
+
+Unfortunately, the conditions for slightly more complex structures
+quickly grow into sequences of partly interdependent bits of
+structural tests, hurting readbility and maintability.  Pattern
+matching offers here a more direct scheme of expressing the minimal
+structure that an objects needs to have in order to allow for further
+processing.  Rather than writing a series of manual tests, patterns
+follow a *declarative* style, which is well-known for improved
+readability, maintability, and for delegating the burden of efficient
+execution on the compiler and interpreter.
+
+The concept of pattern matching is similar to regular expressions,
+where succinct patterns describe a textual structure.  A dedicated
+compiler then transforms these declarative patterns into highly
+efficient finite state machines.  In contrast to regular expressions,
+pattern matching targets Python objects rather than textual data,
+and builds on *decision trees* as the driving motor for finding a
+match.  Moreover, pattern matching blends the matching of a suitable
+pattern with that of a function, i.e. code that is executed in order
+to handle and process the information provided by a specific kind of
+object.
+
+One of the simplest forms of pattern matching that we find in other
+languages comes in the form of *function overloading*.  The type and
+number of arguments then determine which implementation of a specific
+function will be executed.  Object-oriented languages (including
+Python) may also use the *visitor pattern* to differentiate an action
+based on the type or class of an object.  Both of these approaches,
+however, are aimed at 'shallow' structures with little or no direct
+support for nested structures or structural information that is not
+directly encoded in an object's class or type.  For instance, it is
+simple to differentiate between an integer, a string, and a tuple, say,
+but becomes quite cumbersome and difficult to differentiate between
+tuples of different lengths, or between one containing string elements
+vs. one containing numeric elements.  This is where pattern matching
+shines: for structures that go beyond simple class distinctions.
+
+Although pattern matching is a concept that has been known and used
+for decades, we propose a re-interpretation that centres around the
+principle of 'duck typing' and builds on existing features of the
+Python language such as iterable unpacking.  Patterns adopt the syntax
+of parameters as far as possible and, to a somewhat lesser degree,
+that of targets in iterable unpacking.  In contrast to iterable
```

Comment:
I'd start a new paragraph here. Or maybe the reason we don't allow non-sequence iterables should be placed in the rationale for sequence patterns.

Reply:
```suggestion
that of targets in iterable unpacking.

In contrast to iterable
```